_posts/2022-06-13-aws-rds-iam-authentication-spring-boot.md

```diff
@@ -0,0 +1,122 @@
+
+However, if you are deploying your application on the AWS cloud, you don't have to struggle passwords anymore or need a fancy tool, because you simply don't need a password.
+You are now able to establish a database connection by authenticating through IAM. 
+Note that this feature only works for MariaDB, MySQL and PostgreSQL. 
+
+The feature works with "authentication tokens", which is a string of characters that is unique and generated by Amazon RDS.
+One authentication token has a lifespan of 15 minutes and needs to be re-generated before or after it expires to keep the connection alive.
+
+This very useful mechanic can deeply strengthen the security of your application by erasing the need of a password and therefore, destroying another opportunity to get your sensitive data in the wrong hands.
+
+Note that you need to explicitly [enable IAM authentication](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.Enabling.html){:target="_blank" rel="noopener noreferrer"} before you can use this feature.
+
+## Database role grant
+Enabling IAM authentication alone is not enough.
+When you create a new database user for your application, you also need to grant the rights to the database user to fetch an authentication token to use in the authentication process.
+
+For Postgres, this can be done with the following SQL query:
+```sql
+CREATE USER <databaseuser> WITH LOGIN;
+GRANT rds_iam TO databaseuser;
+```
+
+For MySQL:
+```sql
+CREATE USER <databaseuser> IDENTIFIED WITH AWSAuthenticationPlugin AS 'RDS';
+```
+
+## AWS
+The application can only fetch a new authentication token if it has the right to do so.
+Since we are running our application on an AWS environment, it only makes sense to make use of IAM roles.
+
+```json
+{
+   "Version": "2012-10-17",
+   "Statement": [
+      {
+         "Effect": "Allow",
+         "Action": [
+             "rds-db:connect"
+         ],
+         "Resource": [
+             "arn:aws:rds-db:region:account-id:dbuser:DbiResourceId/db-user-name"
+         ]
+      }
+   ]
+}
+```
+
+If you are running your application on an EC2 instance, make sure that your EC2 instance has an IAM role attached to it which explicitly gives access to the `rds-db:connect` action.
+On EKS (Kubernetes for AWS), you can use [IAM roles for service accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html){:target="_blank" rel="noopener noreferrer"}.
+
+
+## Spring Boot
+To achieve the token fetching in Spring Boot, I created a custom `HikariDataSource` which overrides the `getPassword()` method and used the AWS Java SDK to create a request for an authentication token from RDS.
+
+Note that for MariaDB, you do not need your own custom implementation as [they have an implementation of their own](https://mariadb.com/kb/en/mariadb-connector-j-250-release-notes/#aws-iam){:target="_blank" rel="noopener noreferrer"}. 
+
```

Comment:
Maybe you can mention what you need to add for Maven (pom.xml), which dependecy do you need to add for AWS Java SDK, and do you need all of it to work?